distribution/bwc/build.gradle

```diff
@@ -221,8 +221,14 @@ bwcVersions.forPreviousUnreleased { BwcVersions.UnreleasedVersionInfo unreleased
         Task bwcTask = createRunBwcGradleTask(buildBwcTaskName(projectName)) {
             inputs.file("${project.buildDir}/refspec")
             outputs.files(projectArtifact)
-            outputs.cacheIf { true }
+            outputs.cacheIf {
+                // Don't bother caching in 'master' since the BWC branches move too quickly to make this cost worthwhile
+                project.ext.isCi && System.getenv('GIT_BRANCH')?.endsWith("master") == false
```

Comment:
I don't love the fact that we have a CI specific conditional here, but don't have any alternatives to suggest. 
Maybe some generic way to turn of caching for arbitrary tasks and configure that from CI ? 

Also this won't work correctly for PRs that target master, guess that's not as big of an issue because those don't write to the cache ? 

It would be so great if we could come up with a solution that does not involve special casing master.

Reply:
> Also this won't work correctly for PRs that target master, guess that's not as big of an issue because those don't write to the cache ?

Correct, it doesn't really matter. They'll still make cache requests that will miss but that's pretty minimal.

Special logic for "if I'm in CI do this" is really commonplace and I'm honestly surprised we don't have instances of this already. To me, I'd much rather have this kind of thing in the build than in CI config (that is complicated enough) and it also allows us to do this branch-specific (CI config is the same for all branches). To me the more we can push logic out of JJB and into the build we should strive to do that. It's easier to reason about, cuts down on JJB and as mentioned, allows us to do branch-specific logic which we're limited in until we move to JJBB.

My preference was to somehow base this on version information the build had but I don't think that's possible. The `7x` branch has no idea there is another newer major version that exists ahead of it.

Reply:
It's not so much that we do something for CI specifically that bother me, and I hear you very well on the JJb complexity, it's more how hard to find it is, and how hard it will be to reason about if we keep adding these. 
It's fine if we want to configure it in the build but it would be worth to do in a way that produces  an overview of what the differences are. 
For example including a build script conditionally on cI and having all the configuration there with wrapped in  `project(":distribution:bwc") {}`

Reply:
Yeah, there are pros/cons to that approach as well. As if I'm looking at a particular plugin or build script, I also have to consider if there's any CI-specific config in this separate script. This approach would be predicated on use assuming there's CI-specific config at play w/ regards to any particular issue.

Personally, if I'm tracking down a build config issue I'm naturally going to go to the source of that logic. I think I'd rather have any conditionals regarding CI or otherwise to live there. Again, I've dealt with may projects with CI-specific conditionally logic in this way and it's generally not been a problem or source of confusion.

To help out here specifically I've added added a description to this `cacheIf` condition (as we really should do in all cases) which will make this specific condition very obvious in build scans.